src/pip/_internal/req/req_install.py

```diff
@@ -565,31 +577,78 @@ def pyproject_toml(self):
 
         return pp_toml
 
-    def get_pep_518_info(self):
-        """Get PEP 518 build-time requirements.
+    def load_pyproject_toml(self):
+        """Load pyproject.toml.
 
-        Returns the list of the packages required to build the project,
-        specified as per PEP 518 within the package. If `pyproject.toml` is not
-        present, returns None to signify not using the same.
+        We cache the loaded data, so we only load and parse the file once.
+        Also, we extract the two values we care about (requires and
+        build-backend) and discard the rest.
         """
-        # If pyproject.toml does not exist, don't do anything.
-        if not os.path.isfile(self.pyproject_toml):
-            return None
+        if self._pyproject_toml_loaded:
+            return
+
+        # Don't do this processing twice
+        self._pyproject_toml_loaded = True
+
+        has_pyproject = os.path.isfile(self.pyproject_toml)
+        has_setup = os.path.isfile(self.setup_py)
+
+        if has_pyproject:
+            with io.open(self.pyproject_toml, encoding="utf-8") as f:
+                pp_toml = pytoml.load(f)
+            build_system = pp_toml.get("build-system")
+        else:
+            build_system = None
+
+        # The following cases must use PEP 517
+        # We check for use_pep517 equalling False because that
+        # means the user explicitly requested --no-use-pep517
+        if (has_pyproject and not has_setup):
+            if self.use_pep517 is False:
+                raise InstallationError(
+                    "Disabling PEP 517 processing is invalid: "
+                    "project does not have a setup.py"
+                )
+            self.use_pep517 = True
+        if (build_system and "build-backend" in build_system):
+            if self.use_pep517 is False:
+                raise InstallationError(
+                    "Disabling PEP 517 processing is invalid: "
+                    "project specifies a build backend of {} "
+                    "in pyproject.toml".format(
+                        build_system["build-backend"]
+                    )
+                )
+            self.use_pep517 = True
+
+        # Choose whether to use PEP 517 if the user didn't say
+        if self.use_pep517 is None:
+            if has_pyproject:
+                self.use_pep517 = True
+            else:
+                self.use_pep517 = False
+
+        if build_system is None:
+            if self.use_pep517:
+                build_system = {
+                    "requires": ["setuptools>=38.2.5", "wheel"],
```

Comment:
A comment explaining the seemingly magic version number would be nice.

Reply:
Agreed, I've added one.